runtime/src/iree/hal/drivers/hip/cleanup_thread.c

```diff
@@ -0,0 +1,160 @@
+      if (!iree_status_is_ok(status)) {
+        thread->failure_status = status;
+      }
+      if (!iree_status_is_ok(thread->failure_status)) {
+        status = iree_status_clone(thread->failure_status);
+      }
+    }
+    iree_slim_mutex_unlock(&thread->mutex);
+
+    if (!iree_status_is_ok(status) || exit) {
+      break;
+    }
+  }
+  return 0;
+}
+
+iree_status_t iree_hal_hip_cleanup_thread_initialize(
+    const iree_hal_hip_dynamic_symbols_t* symbols,
+    iree_allocator_t host_allocator,
+    iree_hal_hip_cleanup_thread_t** out_thread) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  iree_hal_hip_cleanup_thread_t* thread = NULL;
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0,
+      iree_allocator_malloc(host_allocator, sizeof(*thread), (void**)&thread));
+
+  thread->symbols = symbols;
+  thread->do_exit = false;
+  iree_slim_mutex_initialize(&thread->mutex);
+  iree_hal_hip_callback_queue_initialize(host_allocator, &thread->queue);
+  thread->failure_status = iree_ok_status();
+  thread->host_allocator = host_allocator;
+  iree_notification_initialize(&thread->notification);
+
+  iree_thread_create_params_t params;
+  memset(&params, 0x00, sizeof(params));
+  params.name = iree_make_cstring_view("iree-hal-hip-cleanup");
+  iree_status_t status =
+      iree_thread_create((iree_thread_entry_t)iree_hal_hip_cleanup_thread_main,
+                         thread, params, host_allocator, &thread->thread);
+  if (!iree_status_is_ok(status)) {
+    iree_hal_hip_callback_queue_deinitialize(&thread->queue);
+    iree_slim_mutex_deinitialize(&thread->mutex);
+    iree_allocator_free(host_allocator, thread);
+  }
+  IREE_TRACE_ZONE_END(z0);
+  out_thread[0] = thread;
+  return status;
+}
+
+// Deinitializes the cleanup thread for HIP driver.
+void iree_hal_hip_cleanup_thread_deinitialize(
+    iree_hal_hip_cleanup_thread_t* thread) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_slim_mutex_lock(&thread->mutex);
+  thread->do_exit = true;
+  iree_slim_mutex_unlock(&thread->mutex);
+
+  iree_notification_post(&thread->notification, IREE_ALL_WAITERS);
+  // There is only one owner for the thread, so this also joins the thread.
+  iree_thread_release(thread->thread);
+
+  iree_hal_hip_callback_queue_deinitialize(&thread->queue);
+  iree_slim_mutex_deinitialize(&thread->mutex);
+  iree_allocator_free(thread->host_allocator, thread);
+  IREE_TRACE_ZONE_END(z0);
+}
+
+iree_status_t iree_hal_hip_cleanup_thread_add_cleanup(
+    iree_hal_hip_cleanup_thread_t* thread, iree_hal_hip_event_t* event,
+    iree_hal_hip_cleanup_callback_t callback, void* user_data) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  iree_slim_mutex_lock(&thread->mutex);
+  if (!iree_status_is_ok(thread->failure_status)) {
+    IREE_TRACE_ZONE_END(z0);
+    iree_slim_mutex_unlock(&thread->mutex);
+    return thread->failure_status;
+  }
+
+  iree_hal_hip_cleanup_thread_callback_t callback_data = {
+      .callback = callback, .user_data = user_data, .event = event};
```

Comment:
always include trailing , in designated initializers so clang-format makes them easier to read
```suggestion
  iree_hal_hip_cleanup_thread_callback_t callback_data = {
    .callback = callback,
    .user_data = user_data,
    .event = event,
  };
```